src/Features/Core/Portable/Common/AbstractProjectExtensionProvider.cs

```diff
@@ -29,27 +28,44 @@ internal abstract class AbstractProjectExtensionProvider<TProvider, TExtension,
         protected abstract ImmutableArray<string> GetLanguages(TExportAttribute exportAttribute);
         protected abstract bool TryGetExtensionsFromReference(AnalyzerReference reference, out ImmutableArray<TExtension> extensions);
 
+        public static bool TryGetCachedExtensions(IReadOnlyList<AnalyzerReference> analyzerReferences, out ImmutableArray<TExtension> extensions)
+        {
+            if (s_referencesToExtensionsMap.TryGetValue(analyzerReferences, out var providers))
+            {
+                extensions = providers.Value;
+                return true;
+            }
+
+            extensions = ImmutableArray<TExtension>.Empty;
+            return false;
+        }
+
         public static ImmutableArray<TExtension> GetExtensions(Project? project)
         {
             if (project is null)
                 return ImmutableArray<TExtension>.Empty;
 
-            if (s_referencesToExtensionsMap.TryGetValue(project.AnalyzerReferences, out var providers))
-                return providers.Value;
+            return GetExtensions(project.Language, project.AnalyzerReferences);
+        }
+
+        public static ImmutableArray<TExtension> GetExtensions(string language, IReadOnlyList<AnalyzerReference> analyzerReferences)
+        {
+            if (TryGetCachedExtensions(analyzerReferences, out var providers))
+                return providers;
 
-            return GetExtensionsSlow(project);
+            return GetExtensionsSlow(language, analyzerReferences);
```

Comment:
Any idea why we're using a local function here?

Reply:
Probably to avoid the lambda allocation in slow path when we hit the fast path?